pkg/runtime/controller.go

```diff
@@ -0,0 +1,420 @@
+			}
+		}
+		rule.actions = ruleActions
+		rn[k.Name] = rule
+	}
+
+	return ruleConfig
+}
+
+var cleanupSleepTime = 500 * time.Millisecond
+
+// processActions prunes actions that lack referential integrity and associate instances with
+// handlers that are later used to create new handlers.
+func (c *Controller) processActions(acts []*cpb.Action, handlerConfig map[string]*cpb.Handler,
+	instanceConfig map[string]*cpb.Instance, ht *handlerTable) map[adptTmpl.TemplateVariety]map[string]*Action {
+	actions := make(map[adptTmpl.TemplateVariety]map[string]*Action)
+	for _, ic := range acts {
+		var hc *cpb.Handler
+		if hc = handlerConfig[ic.Handler]; hc == nil {
+			if glog.V(3) {
+				glog.Warningf("ConfigWarning unknown handler: %s", ic.Handler)
+			}
+			continue
+		}
+
+		for _, instName := range ic.Instances {
+			var inst *cpb.Instance
+			if inst = instanceConfig[instName]; inst == nil {
+				if glog.V(3) {
+					glog.Warningf("ConfigWarning unknown instance: %s", instName)
+				}
+				continue
+			}
+
+			ht.Associate(ic.Handler, instName)
+
+			ti := c.templateInfo[inst.Template]
+			// grouped by adapterName.
+			vAction := actions[ti.Variety]
+			if vAction == nil {
+				vAction = make(map[string]*Action)
+				actions[ti.Variety] = vAction
+			}
+
+			act := vAction[hc.Adapter]
+			if act == nil {
+				act = &Action{
+					processor:   &ti,
+					handlerName: ic.Handler,
+					adapterName: hc.Adapter,
+				}
+				vAction[hc.Adapter] = act
+			}
+			act.instanceConfig = append(act.instanceConfig, inst)
+		}
+	}
+	return actions
+}
+
+// combineRulesHandlers sets handler references in rulesConfig.
+// Reject actions from rulesConfig whose handler could not be initialized.
+func combineRulesHandlers(ruleConfig rulesMapByNamespace, handlerTable map[string]*HandlerEntry) {
+	// map by namespace
+	for ns, nsmap := range ruleConfig {
+		// map by rule name
+		for rn, rule := range nsmap {
+			// map by template variety
+			for vr, vact := range rule.actions {
+				newvact := vact[:0]
+				for _, act := range vact {
+					he := handlerTable[act.handlerName]
+					if he == nil {
+						glog.Warningf("Internal error: Handler %s could not be found", act.handlerName)
+						continue
+					}
+					if he.Handler == nil {
+						glog.Warningf("Filtering action from rule %s/%s. Handler %s could not be initialized due to %s.", ns, rn, act.handlerName, he.HandlerCreateError)
+						continue
+					}
+					act.handler = he.Handler
+					newvact = append(newvact, act)
+				}
+				if len(newvact) > 0 {
+					rule.actions[vr] = newvact
+				} else {
+					delete(rule.actions, vr)
+				}
+			}
+			if len(rule.actions) == 0 {
+				glog.Warningf("Purging rule %v with no actions")
+				delete(nsmap, rn)
+			}
+		}
+	}
+}
+
+// cleanupResolver cleans up handler table in the resolver
+// after the resolver is no longer in use.
+func cleanupResolver(r *resolver, table map[string]*HandlerEntry, timeout time.Duration) error {
+	start := time.Now()
+	for {
```

Comment:
We are not really swapping the value here. 
We are polling the refcount in an atomic way until a timeout. This code should only read the value. CAS does not apply here.